### PR TITLE
GSAGH-561 contour legend in rhino is cropped

### PR DESCRIPTION
**Pull request by DominikaLos**
**Scale: 3**
image

**Scale: 1.5**
image

Update: added extraOffset to make semicolon fit in first line

### CODE REVIEW

File: GsaGH/Components/6_Display/AssemblyResults.cs [hunk 1]
```diff
@@ -145,38 +140,20 @@ public override void CreateAttributes() {
...
-          false, (int)(defaultTextHeight * _legendScale));
-      }
-
-      args.Display.Draw2dText(_resType, Color.Black,
-        new Point2d(args.Viewport.Bounds.Right - (int)(110 * _legendScale),
-          (int)(7 * _legendScale)), false, (int)(defaultTextHeight * _legendScale));
-      args.Display.Draw2dText(_case, Color.Black,
-        new Point2d(args.Viewport.Bounds.Right - (int)(110 * _legendScale),
-          (int)(145 * _legendScale)), false, (int)(defaultTextHeight * _legendScale));
+      DrawContour.DrawViewportWires(args, _legend, _resType, _case);
```
Comment (psarras) [inline GsaGH/Components/6_Display/AssemblyResults.cs:143]: I feel the DrawContour is tightly coupled to the Legend, so should we not make a requirement for DrawContour to instantiate using a legend? Or the other way around?

**Review by psarras — CHANGES_REQUESTED** [review]
I tried to do the changes on one of the files, but you can extend them on others.

File: GsaGH/Helpers/GH/Legend/ContourLegendConfiguration.cs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    public Bitmap Bitmap { get; private set; }
+    public double Scale { get; private set; } = 1.0;
+    public bool IsVisible { get; private set; } = true;
+
+    public readonly int DefaultWidth = 15;
+    public readonly int DefaultHeight = 120;
+    /// <summary>
+    ///   Key used to de/serialise scale of the legend
+    /// </summary>
+    public static string ScaleKey => "legendScale";
```
Comment (psarras) [inline GsaGH/Helpers/GH/Legend/ContourLegendConfiguration.cs:21]: these could be const

File: GsaGH/Helpers/GH/Legend/ContourLegendConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    /// <summary>
+    ///   Key used to de/serialise visibility of the legend
+    /// </summary>
+    public static string VisibilityKey => "legend";
```
Comment (psarras) [inline GsaGH/Helpers/GH/Legend/ContourLegendConfiguration.cs:25]: const

File: GsaGH/Helpers/GH/Legend/ContourLegendManager.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    //drawing
+    public void DrawLegend(
+      IGH_PreviewArgs args, string title, string bottomText,
+      List<(int startY, int endY, Color gradientColor)> gradients) {
+      _legend.DrawLegendRectangle(args, title, bottomText, gradients);
+    }
+
+    //menu
+    public ToolStripMenuItem CreateMenu(GH_OasysDropDownComponent component, Action updateUI) {
```
Comment (psarras) [inline GsaGH/Helpers/GH/Legend/ContourLegendManager.cs:41]: they all seem to have the same `Action` ? can't we not set it here without the need of the argument?

File: GsaGH/Helpers/GH/Legend/ContourLegendManager.cs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      return (_configuration.Bitmap.Width, _configuration.Bitmap.Height);
+    }
+
+    public void SetTextValues(List<string> values) {
+      _configuration.SetTextValues(values);
+    }
+
+    public void SetPositionYValues(List<int> values) {
+      _configuration.SetValuePositionsY(values);
+    }
```
Comment (psarras) [inline GsaGH/Helpers/GH/Legend/ContourLegendManager.cs:75]: do we need those pass through methods? Isn't the configuration available to the components?

File: GsaGHTests/Helpers/GH/TextWrapperTests.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+      string text2 = TextWrapper.WrapText(inputText, maxWidth, fontSize);
+
+      Assert.Single(GetPrivateFieldValue());
+    }
+
+    private Dictionary<(string, float), float> GetPrivateFieldValue() {
+      FieldInfo fieldInfo
+        = typeof(TextWrapper).GetField("textWidthCache", BindingFlags.NonPublic | BindingFlags.Static);
+      return (Dictionary<(string, float), float>)fieldInfo.GetValue(null);
+    }
```
Comment (psarras) [inline GsaGHTests/Helpers/GH/TextWrapperTests.cs:48]: reflection here? If we really need to make it `internal` But what are you checking that the cache exists?

As we mentioned on the call, this creates unnecessary complexity, I would suggest remove the cache, if the call is super fash.

File: GsaGHTests/Helpers/GH/TextWrapperTests.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+      string inputText = "This is a test sentence that should be wrapped correctly.";
+      int maxWidth = 100;
+      int fontSize = 10;
+
+      string result = TextWrapper.WrapText(inputText, maxWidth, fontSize);
+
+      Assert.NotNull(result);
+      Assert.Contains("\n", result);
+      Assert.Equal("This is a test\nsentence that\nshould be\nwrapped\ncorrectly.", result);
+    }
```
Comment (psarras) [inline GsaGHTests/Helpers/GH/TextWrapperTests.cs:30]: how did you figure what is the right result? I bet you had to run it once and whatever came out you use as "expected". 

I would have a test, where you put a really small `Width` let's say `1` and then put a text: `12` and then you expect to get `1\n2` which is understandable.

Then if you can make a more complex, where we can specify the length of the word and the length of the sentence by Mocking the algorithm then that is fine but it is secondary.

File: GsaGHTests/Helpers/GH/TextWrapperTests.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+      FieldInfo fieldInfo
+        = typeof(TextWrapper).GetField("textWidthCache", BindingFlags.NonPublic | BindingFlags.Static);
+      return (Dictionary<(string, float), float>)fieldInfo.GetValue(null);
+    }
+
+    [Fact]
+    public void GetFontName_ShouldReturnDefaultWhenRhinoDocIsNull() {
+      string fontName = TextWrapper.GetFontName();
+      Assert.Equal("Arial", fontName);
+    }
```
Comment (psarras) [inline GsaGHTests/Helpers/GH/TextWrapperTests.cs:54]: So in what situation the user will get "Arial" font on their legend? It feels this is not something we want to specify? Maybe I am missing something.

File: GsaGHTests/Helpers/GH/Legend/LegendMenuBuilderTests.cs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    [Fact]
+    public void CreateLegendToolStripMenuItem_UpdatesTextBox_OnTextChange() {
+      ToolStripItem scaleTextBox = _menuItem.DropDownItems[0];
+      double newScale = 2.5d;
+      // Act - Simulate text change
+      scaleTextBox.Text = $"{newScale}"; // Updates private _scaleLegendTxt indirectly
+      _menuItem.DropDownItems[1].PerformClick(); // Simulate mouse up to trigger UpdateLegendScale
+
+      Assert.Equal(newScale.ToString(), _menuItem.DropDownItems[0]?.Text);
+    }
```
Comment (psarras) [inline GsaGHTests/Helpers/GH/Legend/LegendMenuBuilderTests.cs:47]: is this unfinished? I don't understand what it is testing on line 43 you are updating the value, and then on the assert you are checking if it got updated? :)

File: GsaGHTests/Helpers/GH/Legend/LegendMenuBuilderTests.cs [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+namespace GsaGHTests.Helpers {
+  [Collection("GrasshopperFixture collection")]
+  public class LegendMenuBuilderTests {
+    private Mock<Contour1dResults> _componentMock;
+    private double _initialScale;
+    private ToolStripMenuItem _menuItem;
+
+    public LegendMenuBuilderTests() {
+      _componentMock = new Mock<Contour1dResults>(); // any component you want
+      var updateUiMock = new Mock<Action>();
```
Comment (psarras) [inline GsaGHTests/Helpers/GH/Legend/LegendMenuBuilderTests.cs:20]: do we need to mock the update? I mean you will remove the UpdateUi either way but this one could have been: `() => {}`

File: GsaGHTests/Helpers/GH/Legend/ContourLegendTests.cs [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        for (int x = 0; x < _mockBitmap.Width; x++) {
+          Assert.Equal(color, _mockBitmap.GetPixel(x, _mockBitmap.Height - y - 1));
+        }
+      }
+    }
+
+    [Fact]
+    public void DrawGradientWillThrowErrorWhenStartValuesIsBiggerThanEndValue() {
+      Color color = Color.Red;
+      Assert.Throws<ArgumentOutOfRangeException>(() => legend.DrawGradientLegend(5, 1, color));
```
Comment (psarras) [inline GsaGHTests/Helpers/GH/Legend/ContourLegendTests.cs:49]: minor thing, but use the least amount of values. 

```suggestion
      Assert.Throws<ArgumentOutOfRangeException>(() => legend.DrawGradientLegend(2, 1, color));
```

File: GsaGHTests/Helpers/GH/Legend/ContourLegendTests.cs [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+namespace GsaGHTests.Helpers {
+  [Collection("GrasshopperFixture collection")]
+  public class ContourLegendTests {
+    private readonly Mock<IContourLegendConfiguration> _mockConfiguration;
+    private ContourLegend legend;
+    private readonly Bitmap _mockBitmap;
+
+    public ContourLegendTests() {
+      _mockBitmap = new Bitmap(10, 10);
```
Comment (psarras) [inline GsaGHTests/Helpers/GH/Legend/ContourLegendTests.cs:22]:
```suggestion
      _bitmap = new Bitmap(10, 10);
```

File: GsaGHTests/Helpers/GH/Legend/ContourLegendTests.cs [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+    [Fact]
+    public void DrawGradientWillThrowErrorWhenStartValuesIsBiggerThanEndValue() {
+      Color color = Color.Red;
+      Assert.Throws<ArgumentOutOfRangeException>(() => legend.DrawGradientLegend(5, 1, color));
+    }
+
+    [Fact]
+    public void DrawGradientWillThrowErrorWhenEndValueIsBiggerThanBitmapHeight() {
+      Color color = Color.Red;
+      Assert.Throws<ArgumentOutOfRangeException>(() => legend.DrawGradientLegend(5, 200, color));
```
Comment (psarras) [inline GsaGHTests/Helpers/GH/Legend/ContourLegendTests.cs:55]:
```suggestion
      Assert.Throws<ArgumentOutOfRangeException>(() => legend.DrawGradientLegend(1, _bitmap.Height + 1, color));
```

File: GsaGH/Components/6_Display/Contour2dResults.cs
Comment (psarras) [inline GsaGH/Components/6_Display/Contour2dResults.cs:1]: typos

File: GsaGH/Components/6_Display/Contour3dResults.cs
Comment (psarras) [inline GsaGH/Components/6_Display/Contour3dResults.cs:1]: typos

File: GsaGH/Components/6_Display/ContourNodeResults.cs
Comment (psarras) [inline GsaGH/Components/6_Display/ContourNodeResults.cs:1]: typos

File: GsaGH/Helpers/GH/Legend/ContourLegendConfiguration.cs [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+        }
+
+        _width = value;
+      }
+    }
+    public int Height {
+      get => _height;
+      set {
+        if (value <= 0) {
+          throw new ArgumentOutOfRangeException(nameof(_height), "Height must be greater than zero.");
```
Comment (psarras) [inline GsaGH/Helpers/GH/Legend/ContourLegendConfiguration.cs:52]: same as before `nameof(Height, Width etc...)

**Review by psarras — CHANGES_REQUESTED** [review]
resolve the comments you caught up with already

File: GsaGH/Helpers/GH/Legend/LegendMenuBuilder.cs [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+      legendScaleTextBox.TextChanged += (s, e) => MaintainScaleLegendText(legendScaleTextBox);
+
+      var legendScaleMenuItem = new ToolStripMenuItem("Scale Legend") {
+        Enabled = true,
+        ImageScaling = ToolStripItemImageScaling.SizeToFit,
+      };
+
+      var menuControl
+        = new GH_MenuCustomControl(legendScaleMenuItem.DropDown, legendScaleTextBox.Control, true,
+          200); // needed! don't remove
```
Comment (psarras) [inline GsaGH/Helpers/GH/Legend/LegendMenuBuilder.cs:34]:
```suggestion
      _ = new GH_MenuCustomControl(legendScaleMenuItem.DropDown, legendScaleTextBox.Control, true,
        200);
```

File: GsaGH/Helpers/GH/TextWrapper.cs [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    /// <param name="maxWidth">The maximum width in pixels.</param>
+    /// <param name="font">The system font</param>
+    /// <returns>Wrapped text with line breaks.</returns>
+    public static string WrapText(string text, int maxWidth, Font font) {
+      if (string.IsNullOrWhiteSpace(text)) {
+        return string.Empty;
+      }
+
+      string[] words = GetTextToWrap(text, out List<string> lines);
+      string currentLine = "";
```
Comment (psarras) [inline GsaGH/Helpers/GH/TextWrapper.cs:20]:
```suggestion
      string currentLine = string.Empty;
```

File: GsaGH/Helpers/GH/TextWrapper.cs [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+        textToWrap = splittedText[1];
+      } else {
+        textToWrap = text;
+      }
+
+      return textToWrap.Split(' ');
+    }
+
+    /// <summary>
+    ///   Gets the width of the given text from the cache or measures and caches it.
```
Comment (psarras) [inline GsaGH/Helpers/GH/TextWrapper.cs:56]: this comment is not longer true. let's remove the comments, it is pretty self explanatory

File: GsaGH/Helpers/GH/TextWrapper.cs [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+    /// </summary>
+    /// <param name="text">The text to measure.</param>
+    /// <param name="font">Font used for rendering text</param>
+    /// <param name="maxWidth">Maximum possible width for text</param>
+    /// <returns>The width of the text in pixels.</returns>
+    private static float CalculateTextWidth(string text, Font font, int maxWidth) {
+      var graphics = System.Drawing.Graphics.FromImage(new Bitmap(maxWidth, 1)); //we care only about width
+      return graphics.MeasureString(text, font).Width;
```
Comment (psarras) [inline GsaGH/Helpers/GH/TextWrapper.cs:64]: so I think this one doesn't have to specify the maxWidth. I've created an example with a new Bitmap(1, 1) and the MeasureString was able to return more than 1 length

```csharp
    [Fact]
    public void ShouldMeasureLongerThanBitmap() {
      var font = new Font(SystemFonts.DefaultFont.FontFamily, SystemFonts.DefaultFont.Size);
      Graphics graphics = Graphics.FromImage(new Bitmap(1, 1)); //we care only about width
      var a = graphics.MeasureString("a really long text", font).Width;
      Assert.True(a > 1);
      Assert.Equal(0, a);
    }
```

In that case I am suggesting to create a Graphics object and cache it on the class itself and reuse it all the time to measure the text.

```suggestion
    private static Bitmap bitmap = new Bitmap(1, 1);
    private static Graphics graphics = Graphics.FromImage(bitmap);


    private static float CalculateTextWidth(string text, Font font, int maxWidth) {
      return graphics.MeasureString(text, font).Width;
```

File: GsaGHTests/3_Components/6_Display/Contour2dResultsTests.cs [hunk 19]
```diff
@@ -171,13 +171,6 @@ public void UpdateStressTest() {
       comp.UpdateStress("MPa");
     }
 
-    [Fact]
-    public void UpdateLegendScaleTest() {
-      var comp = new Contour2dResults();
-      ComponentTestHelper.SetInput(comp, GsaResultTests.NodeAndElement2dCombinationResultsMother());
-      comp.UpdateLegendScale();
-    }
```
Comment (psarras) [inline GsaGHTests/3_Components/6_Display/Contour2dResultsTests.cs:179]: I see no assert here, but why did we remove it? Looks like a clean-up, but was it capturing anything?

**Review by psarras — CHANGES_REQUESTED** [review]

File: GsaGHTests/3_Components/6_Display/Contour2dResultsTests.cs (hunk 19 above)
Comment (DominikaLos) [reply GsaGHTests/3_Components/6_Display/Contour2dResultsTests.cs:179]: there is no more updateLegendScale method in contour components. no needed

**Review by psarras — APPROVED** [review]

**Pull request merged**